cpp/src/arrow/compute/function.cc

```diff
@@ -167,6 +180,109 @@ const Kernel* DispatchExactImpl(const Function* func,
   return nullptr;
 }
 
+struct FunctionExecutorImpl : public FunctionExecutor {
+  FunctionExecutorImpl(std::vector<TypeHolder> in_types, const Kernel* kernel,
+                       std::unique_ptr<detail::KernelExecutor> executor,
+                       const Function& func)
+      : in_types(std::move(in_types)),
+        kernel(kernel),
+        kernel_ctx(),
+        executor(std::move(executor)),
+        func(func),
+        state(),
+        options(NULLPTR) {}
+  virtual ~FunctionExecutorImpl() {}
+
+  Status KernelInit(const FunctionOptions* options) {
+    if (!kernel_ctx) {
```

Comment:
KernelContext is a simple wrapper around an ExecContext, KernelState, and a Kernel. I think it's not worthwhile to keep it by pointer and it'd be preferable validate the members of the KernelContext instead- like checking if KernelContext::exec_context is non null (or coercing to default_exec_context)

Reply:
Will do.